Source/Core/VideoCommon/UberShaderVertex.cpp

```diff
@@ -0,0 +1,577 @@
+{
+  const bool msaa = g_ActiveConfig.iMultisamples > 1;
+  const bool ssaa = g_ActiveConfig.iMultisamples > 1 && g_ActiveConfig.bSSAA;
+  const u32 numTexgen = uid_data->num_texgens;
+  ShaderCode out;
+
+  out.Write("// Vertex UberShader\n\n");
+  WriteUberShaderCommonHeader(out, ApiType);
+
+  out.Write("%s", s_lighting_struct);
+
+  // uniforms
+  if (ApiType == APIType::OpenGL || ApiType == APIType::Vulkan)
+    out.Write("UBO_BINDING(std140, 2) uniform VSBlock {\n");
+  else
+    out.Write("cbuffer VSBlock {\n");
+  out.Write(s_shader_uniforms);
+  out.Write("};\n");
+
+  out.Write("int4 CalculateLighting(uint index, uint attnfunc, uint diffusefunc, float4 pos, "
+            "float3 _norm0) {\n"
+            "  float3 ldir, h, cosAttn, distAttn;\n"
+            "  float dist, dist2, attn;\n"
+            "\n"
+            "  switch (attnfunc) {\n");
+  out.Write("  case %uu: // LIGNTATTN_NONE\n", LIGHTATTN_NONE);
+  out.Write("  case %uu: // LIGHTATTN_DIR\n", LIGHTATTN_DIR);
+  out.Write("    ldir = normalize(" I_LIGHTS "[index].pos.xyz - pos.xyz);\n"
+            "    attn = 1.0;\n"
+            "    if (length(ldir) == 0.0)\n"
+            "      ldir = _norm0;\n"
+            "    break;\n\n");
+  out.Write("  case %uu: // LIGHTATTN_SPEC\n", LIGHTATTN_SPEC);
+  out.Write("    ldir = normalize(" I_LIGHTS "[index].pos.xyz - pos.xyz);\n"
+            "    attn = (dot(_norm0, ldir) >= 0.0) ? max(0.0, dot(_norm0, " I_LIGHTS
+            "[index].dir.xyz)) : 0.0;\n"
+            "    cosAttn = " I_LIGHTS "[index].cosatt.xyz;\n");
+  out.Write("    if (diffusefunc == %uu) // LIGHTDIF_NONE\n", LIGHTDIF_NONE);
+  out.Write("      distAttn = " I_LIGHTS "[index].distatt.xyz;\n"
+            "    else\n"
+            "      distAttn = normalize(" I_LIGHTS "[index].distatt.xyz);\n"
+            "    attn = max(0.0, dot(cosAttn, float3(1.0, attn, attn*attn))) / dot(distAttn, "
+            "float3(1.0, attn, attn*attn));\n"
+            "    break;\n\n");
+  out.Write("  case %uu: // LIGHTATTN_SPOT\n", LIGHTATTN_SPOT);
+  out.Write("    ldir = " I_LIGHTS "[index].pos.xyz - pos.xyz;\n"
+            "    dist2 = dot(ldir, ldir);\n"
+            "    dist = sqrt(dist2);\n"
+            "    ldir = ldir / dist;\n"
+            "    attn = max(0.0, dot(ldir, " I_LIGHTS "[index].dir.xyz));\n"
+            "    attn = max(0.0, " I_LIGHTS "[index].cosatt.x + " I_LIGHTS
+            "[index].cosatt.y * attn + " I_LIGHTS "[index].cosatt.z * attn * attn) / dot(" I_LIGHTS
+            "[index].distatt.xyz, float3(1.0, dist, dist2));\n"
+            "    break;\n\n");
+  out.Write("  default:\n"
+            "    attn = 1.0;\n"
+            "    ldir = _norm0;\n"
+            "    break;\n"
+            "  }\n"
+            "\n"
+            "  switch (diffusefunc) {\n");
+  out.Write("  case %uu: // LIGHTDIF_NONE\n", LIGHTDIF_NONE);
+  out.Write("    return int4(round(attn * " I_LIGHTS "[index].color));\n\n");
```

Comment:
In GLSL-ES (float) * (int4) isn't allowed without explicit conversion to vec4 first